src/protocol/sort/generate_permutation_opt.rs

```diff
@@ -113,6 +117,104 @@ where
     Ok(composed_less_significant_bits_permutation)
 }
 
+#[allow(dead_code)]
+#[embed_doc_image("malicious_sort", "images/sort/malicious-sort.png")]
+/// Returns a sort permutation in a malicious context.
+/// This runs sort in a malicious context. The caller is responsible to validate the accumulater contents and downgrade context to Semi-honest before calling this function
+/// The function takes care of upgrading and validating while the sort protocol runs.
+/// It then returns a semi honest context with output in Replicated format. The caller should then upgrade the output and context before moving forward
+///
+/// Steps
+/// 1. [Malicious Special] Upgrade the context from semihonest to malicious and get a validator
+/// 2. [Malicious Special] Upgrade 0..`num_multi_bits` sort bit keys
+/// 3. Compute bit permutation that sorts 0..`num_multi_bits` bit
+///
+/// For `num_multi_bits` to N-1th bit of input share
+/// 1. i. Shuffle the i-1th composition
+///   ii. [Malicious Special] Validate the accumulator contents
+///  iii. [Malicious Special] Malicious reveal
+///   iv. [Malicious Special] Downgrade context to semihonest
+/// 2. i. [Malicious Special] Upgrade ith sort bit keys
+///   ii. Sort i..i+`num_multi_bits` bits based on i-1th bits by applying i-1th composition on i..i+`num_multi_bits` bits
+/// 3. Compute bit permutation that sorts i..i+`num_multi_bits` bits
+/// 4. Compute ith composition by composing i-1th composition on ith permutation
+/// In the end, following is returned
+///    i. n-1th composition: This is the permutation which sorts the inputs
+///   ii. Validator which can be used to validate the leftover items in the accumulator
+///
+/// ![Malicious sort permutation steps][malicious_sort]
+/// # Panics
+/// If sort keys dont have num of bits same as `num_bits`
+/// # Errors
+pub async fn malicious_generate_permutation_opt<'a, F>(
+    sh_ctx: SemiHonestContext<'a, F>,
+    sort_keys: &[Vec<Vec<Replicated<F>>>],
```

Comment:
This type signature is unhelpful. It doesn't inform the reader that we are expecting a highly specific format.

I'd really like to get some kind of compile-time assistance on this one. Could we somehow make a few types that at compile-time are known to be of specific lengths? For example, a "chunk" could internally be implemented as a vector of arrays of length 3, if the compiler just knew that the number of multi-bits was 3. That would also make for a much more informative type. I'm not sure exactly how to achieve this, but imagine if this was something like our BitArray40 type, where the size is known at compile time. Then we could have something like:

```
pub async fn malicious_generate_permutation_opt<'a, F, Ch>(
    sh_ctx: SemiHonestContext<'a, F>,
    sort_keys: &[Vec<Ch>],
) -> blah blah
where
    F: Field,
    Ch: MultiBitChunk<F>,
{
```

Where we could have various types that all implement `MultiBitChunk`, one of which would be like `ThreeBitChunk` or something.

Reply:
Actually the chunks are all same size apart from the last one. e.g. if num_bits = 40 and multi bits = 3, then we have 13 * [;3] and 1 [;1]. So we would need to handle that scenario as well. Lets work on this in follow up PRs?